pkg/kv/kvserver/rangefeed/event_size_test.go

```diff
@@ -0,0 +1,639 @@
+	case "update_intent":
+		op, futureEvent, mem, futureMem = updateIntentOpEvent(data.txnID, data.timestamp)
+	case "commit_intent":
+		op, futureEvent, mem, futureMem = commitIntentOpEvent(data.txnID, data.key, data.timestamp, data.value, nil, data.omitInRangefeeds)
+	case "abort_intent":
+		op, futureEvent, mem, futureMem = abortIntentOpEvent(data.txnID)
+	case "abort_txn":
+		op, futureEvent, mem, futureMem = abortTxnOpEvent(data.txnID)
+	}
+
+	ev = event{
+		ops: []enginepb.MVCCLogicalOp{op},
+	}
+	expectedMemUsage += eventOverhead + mem
+	expectedFutureMemUsage += futureMem
+	// Publish checkpoint event.
+	expectedFutureMemUsage += futureEventBaseOverhead + rangefeedCheckpointOverhead
+	return ev, futureEvent, expectedMemUsage, expectedFutureMemUsage
+}
+
+func generateCtEvent(
+	data testData,
+) (
+	ev event,
+	futureEvent kvpb.RangeFeedEvent,
+	expectedMemUsage int64,
+	expectedFutureMemUsage int64,
+) {
+	ev = event{
+		ct: ctEvent{
+			Timestamp: data.timestamp,
+		},
+	}
+	expectedMemUsage += eventOverhead
+	// Publish checkpoint event.
+	expectedFutureMemUsage += futureEventBaseOverhead + rangefeedCheckpointOverhead
+	return ev, futureEvent, expectedMemUsage, expectedFutureMemUsage
+}
+
+func generateInitRTSEvent() (
+	ev event,
+	futureEvent kvpb.RangeFeedEvent,
+	expectedMemUsage int64,
+	expectedFutureMemUsage int64,
+) {
+	ev = event{
+		initRTS: true,
+	}
+	expectedMemUsage += eventOverhead
+	// Publish checkpoint event.
+	expectedFutureMemUsage += futureEventBaseOverhead + rangefeedCheckpointOverhead
+	return ev, futureEvent, expectedMemUsage, expectedFutureMemUsage
+}
+
+func generateSSTEvent(
+	t *testing.T, data testData, st *cluster.Settings,
+) (
+	ev event,
+	futureEvent kvpb.RangeFeedEvent,
+	expectedMemUsage int64,
+	expectedFutureMemUsage int64,
+) {
+	sst, _, _ := storageutils.MakeSST(t, st, data.kvs)
+	ev = event{
+		sst: &sstEvent{
+			data: sst,
+			span: data.span,
+			ts:   data.timestamp,
+		},
+	}
+
+	futureEvent.MustSetValue(&kvpb.RangeFeedSSTable{
+		Data:    sst,
+		Span:    data.span,
+		WriteTS: data.timestamp,
+	})
+	expectedMemUsage += eventOverhead + sstEventOverhead + int64(cap(sst)+cap(data.span.Key)+cap(data.span.EndKey))
+	expectedFutureMemUsage += futureEventBaseOverhead + rangefeedSSTTableOverhead + int64(cap(sst)+cap(data.span.Key)+cap(data.span.EndKey))
+	return ev, futureEvent, expectedMemUsage, expectedFutureMemUsage
+}
+
+func generateSyncEvent() (
+	ev event,
+	futureEvent kvpb.RangeFeedEvent,
+	expectedMemUsage int64,
+	expectedFutureMemUsage int64,
+) {
+	ev = event{
+		sync: &syncEvent{c: make(chan struct{})},
+	}
+	expectedMemUsage += eventOverhead + syncEventOverhead
+	return
+}
+
+func generateRandomizedEventAndSend(
+	rand *rand.Rand,
+) (
+	ev event,
+	expectedMemUsage int64,
+	expectedFutureMemUsage int64,
+	randomlyChosenEvent string,
+	randomlyChosenLogicalOp string,
+) {
+	// Opt out sst event since it requires testing.T to call
```

Comment:
The benchmark randomly generates lots of events, and every types of events with equal probability which is not true in production (for example, initRTS should happen much less frequent than logicalOps). I'm will improve the benchmark again with a weighted rand distribution soon.